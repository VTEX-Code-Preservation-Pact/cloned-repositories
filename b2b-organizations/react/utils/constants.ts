export const B2B_CHECKOUT_SESSION_KEY = 'b2b-checkout-settings'
