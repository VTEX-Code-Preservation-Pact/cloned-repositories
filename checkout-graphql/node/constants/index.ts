export enum AddressType {
  RESIDENTIAL = 'residential',
  COMMERCIAL = 'commercial',
  INSTORE = 'inStore',
  GIFT_REGISTRY = 'giftRegistry',
  PICKUP = 'pickup',
  SEARCH = 'search',
}

// Cookies
export const CHECKOUT_COOKIE = 'checkout.vtex.com'
export const ASPXAUTH_COOKIE = '.ASPXAUTH'
export const OWNERSHIP_COOKIE = 'CheckoutOrderFormOwnership'

// Delivery channels
export const DELIVERY = 'delivery'
export const PICKUP_IN_POINT = 'pickup-in-point'

export const VTEX_SESSION = 'vtex_session'
