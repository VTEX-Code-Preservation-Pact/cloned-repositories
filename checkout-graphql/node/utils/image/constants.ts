export const DEFAULT_WIDTH = 96
export const DEFAULT_HDF = 1
