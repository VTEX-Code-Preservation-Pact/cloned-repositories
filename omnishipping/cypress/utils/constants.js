export const BASE_URL = 'http://vtexgame1.vtexlocal.com.br:3000/'
export const BASE_URL_COMPLETE = 'http://vtexgame1.vtexlocal.com.br/'
export const ADD_SKUS_ENDPOINT = 'checkout/cart/add?'
export const CHECKOUT_ENDPOINT = 'checkout/'
export const PROFILE_ENDPOINT = 'api/checkout/pub/profiles/'
