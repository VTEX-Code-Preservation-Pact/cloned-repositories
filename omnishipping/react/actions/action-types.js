export const CHANGE_ACTIVE_TAB = 'CHANGE_ACTIVE_TAB'
export const CHANGE_ACTIVE_SLA_OPTION = 'CHANGE_ACTIVE_SLA_OPTION'
export const SELECT_DELIVERY_WINDOW = 'SELECT_DELIVERY_WINDOW'

export const CHANGE_ACTIVE_PICKUP_OPTIONS = 'CHANGE_ACTIVE_PICKUP_OPTIONS'
export const CHANGE_ACTIVE_PICKUP_SELLER = 'CHANGE_ACTIVE_PICKUP_SELLER'
export const CHANGE_ACTIVE_DELIVERY_CHANNEL = 'CHANGE_ACTIVE_DELIVERY_CHANNEL'
export const CLOSE_PICKUP_MODAL = 'CLOSE_PICKUP_MODAL'
export const OPEN_PICKUP_MODAL = 'OPEN_PICKUP_MODAL'

export const CHANGE_ACTIVE_PICKUP_DETAILS = 'CHANGE_ACTIVE_PICKUP_DETAILS'
export const CLOSE_PICKUP_DETAILS = 'CLOSE_PICKUP_DETAILS'

export const UPDATE_ADDRESS_FORM = 'UPDATE_ADDRESS_FORM'
export const UPDATE_ADDRESS_RULES = 'UPDATE_ADDRESS_RULES'
export const CHANGE_SELECTED_ADDRESS_RULES = 'CHANGE_SELECTED_ADDRESS_RULES'
export const VALIDATE_ADDRESS_FORM = 'VALIDATE_ADDRESS_FORM'
export const VALIDATE_ADDRESS_FORM_FIELDS = 'VALIDATE_ADDRESS_FORM_FIELDS'
export const VALIDATE_POSTAL_CODE = 'VALIDATE_POSTAL_CODE'
export const ADD_ADDRESS_FORM_ADDRESS = 'ADD_ADDRESS_FORM_ADDRESS'
export const ADD_DELIVERY_ADDRESS = 'ADD_DELIVERY_ADDRESS'
export const UPDATE_ORDERFORM = 'UPDATE_ORDERFORM'

export const SELECT_ADDRESS_ITEM = 'SELECT_ADDRESS_ITEM'

export const EDIT_ADDRESS = 'EDIT_ADDRESS'
export const EDIT_OMNISHIPPING = 'EDIT_OMNISHIPPING'
export const NEW_ADDRESS = 'NEW_ADDRESS'
export const SHOW_ADDRESS_PRODUCTS = 'SHOW_ADDRESS_PRODUCTS'
export const SHOW_ADDRESS = 'SHOW_ADDRESS'
export const SHOW_SPLIT = 'SHOW_SPLIT'
export const SHOW_ADDRESS_LIST = 'SHOW_ADDRESS_LIST'
export const SHOW_PAYMENT_BUTTON = 'SHOW_PAYMENT_BUTTON'
export const SHOW_OMNISHIPPING_COMPLETED = 'SHOW_OMNISHIPPING_COMPLETED'
export const SHOW_OMNISHIPPING = 'SHOW_OMNISHIPPING'
export const ENABLE_OMNISHIPPING = 'ENABLE_OMNISHIPPING'
export const DISABLE_OMNISHIPPING = 'DISABLE_OMNISHIPPING'

export const UPDATE_SHIPPING_FAILURE = 'UPDATE_SHIPPING_FAILURE'
export const UPDATE_SHIPPING_FROM_SHIPPING_RATE_PREVIEW =
  'UPDATE_SHIPPING_FROM_SHIPPING_RATE_PREVIEW'
export const UPDATE_SHIPPING_REQUEST = 'UPDATE_SHIPPING_REQUEST'

export const SIMULATE_SHIPPING_REQUEST = 'SIMULATE_SHIPPING_REQUEST'
export const SIMULATE_SHIPPING_FAILURE = 'SIMULATE_SHIPPING_FAILURE'
export const SIMULATE_SHIPPING_SUCCESS = 'SIMULATE_SHIPPING_SUCCESS'

export const UPDATE_SIMULATION_OPTIONS = 'UPDATE_SIMULATION_OPTIONS'

export const REMOVE_ITEMS_REQUEST = 'REMOVE_ITEMS_REQUEST'

export const FOCUS_ON_INPUT = 'FOCUS_ON_INPUT'

export const HAS_PICKUP_OPTIONS_EVENT = 'HAS_PICKUP_OPTIONS_EVENT'
export const SELECTED_PICKUP_TAB_EVENT = 'SELECTED_PICKUP_TAB_EVENT'
export const SHIPPING_STATS_EVENT = 'SHIPPING_STATS_EVENT'
export const SEARCHED_PICKUP_ADDRESS_EVENT = 'SEARCHED_PICKUP_ADDRESS_EVENT'
export const SEARCHED_PICKUP_ADDRESS_GEOLOCATION_EVENT =
  'SEARCHED_PICKUP_ADDRESS_GEOLOCATION_EVENT'

export const VTEX_ID_STARTED = 'VTEX_ID_STARTED'
export const VTEX_ID_CLOSED = 'VTEX_ID_CLOSED'

export const CHANGE_ACTIVE_DELIVERY_PACKAGE = 'CHANGE_ACTIVE_DELIVERY_PACKAGE'

export const UPDATE_SELECTED_PACKAGE_SLIDER = 'UPDATE_SELECTED_PACKAGE_SLIDER'

export const HAS_GEOLOCATION = 'HAS_GEOLOCATION'

export const SELECT_SCHEDULED_DELIVERY = 'SELECT_SCHEDULED_DELIVERY'
export const TOGGLE_SCHEDULED_DELIVERY = 'TOGGLE_SCHEDULED_DELIVERY'

export const OMNISHIPPING_JS_ERROR = 'OMNISHIPPING_JS_ERROR'
export const OMNISHIPPING_API_ERROR = 'OMNISHIPPING_API_ERROR'

export const SET_ADITIONAL_SHIPPING_DATA = 'SET_ADITIONAL_SHIPPING_DATA'

export const OMNISHIPPING_PAGEVIEW_EVENT = 'OMNISHIPPING_PAGEVIEW_EVENT'

export const COMPLETE_OMNISHIPPING = 'COMPLETE_OMNISHIPPING'
