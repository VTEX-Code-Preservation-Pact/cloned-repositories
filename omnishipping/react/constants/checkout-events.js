export const ENABLE = 'enable.vtex'
export const DISABLE = 'disable.vtex'
export const COMPONENT_VALIDATED = 'componentValidated.vtex'
export const COMPONENT_DONE = 'componentDone.vtex'
export const SHIPPING_INFO_UPDATED = 'shippingInformationUpdated.vtex'
export const ORDER_FORM_UPDATED = 'orderFormUpdated.vtex'
export const SHOW_MESSAGE = 'showMessage.vtex'
export const DELIVERY_SELECTED = 'deliverySelected.vtex'
export const AUTHENTICATED_USER = 'authenticatedUser.vtexid'
export const CLOSED_VTEXID = 'closed.vtexid'
