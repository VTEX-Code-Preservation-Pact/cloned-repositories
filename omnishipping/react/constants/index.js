export const PICKUP_IN_STORE = 'pickup-in-point'
export const DELIVERY = 'delivery'
export const PICKUP = 'pickup'
export const SCHEDULED_DELIVERY = 'Entrega Agendada'

export const RESIDENTIAL = 'residential'
export const COMMERCIAL = 'commercial'
export const SEARCH = 'search'
export const GIFT_REGISTRY = 'giftRegistry'

export const GEOCORDS = 'geoCoords'
export const ACCOUNT_INFO_ENDOPOINT =
  '/api/logistics/pub/shipping/configuration'
export const HIDE_MAP = 'HIDE_MAP'
export const SHOW_MAP = 'SHOW_MAP'
export const AVAILABLE = 'available'
export const SHIPPING_HASH = '#/shipping'
export const PAYMENT_HASH = '#/payment'

export const CHEAPEST = 'CHEAPEST'
export const FASTEST = 'FASTEST'
export const COMBINED = 'COMBINED'

export const DENIED = 'denied'

export const VTEXLOCAL = 'vtexlocal'
export const VTEXBETA = 'vtexcommercebeta'
export const VTEXSTABLE = 'vtexcommercestable'

export const DEFAULT_WIDTH = 50
export const DEFAULT_HEIGHT = 50

export const IMPORTANT = 'Important'
export const CRITICAL = 'Critical'
export const ERROR = 'Error'
export const INFO = 'Info'

export const OMNISHIPPING = 'omnishipping'
