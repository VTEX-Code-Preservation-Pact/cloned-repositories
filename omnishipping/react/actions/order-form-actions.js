import * as types from './action-types'
import get from 'lodash/get'
import isEqual from 'lodash/isEqual'
import omit from 'lodash/omit'
import { CHEAPEST, COMBINED, FASTEST } from '../constants'
import { filterSelectedAddresses } from '../utils/OrderFormUtils'
import { isPickup, isDelivery } from '../utils/DeliveryChannelsUtils'
import { getOrderFormItems } from '../selectors/order-form-selectors'
import { omnishippingAPIErrorEvent } from './metric-actions'
import { getSelectedLeanOption } from '../selectors/delivery-selectors'
import { setAditionalShippingData } from './component-activity-actions'
import { validatePostalCode, validateAddressForm } from './address-form-actions'
import {
  getAddress,
  getSearchAddress,
} from '../selectors/address-form-selectors'
import { sortSlasById, hasSLAs } from '../utils/SlasUtils'
import { safelyParseJSON, setItem } from '../utils/LocalStorageUtils'
import { helpers } from 'vtex.address-form'
const { removeValidation } = helpers

const $ = window.$

// ACTIONS
export const changeActiveTabAction = ({ channel, address, searchAddress }) => ({
  type: types.CHANGE_ACTIVE_TAB,
  channel,
  address,
  searchAddress,
})

export const selectScheduledDelivery = isScheduledDeliveryActive => ({
  type: types.SELECT_SCHEDULED_DELIVERY,
  isScheduledDeliveryActive,
})

export const changeActiveSLAOptionAction = ({
  slaOption,
  logisticsInfoIndex,
  sellerId,
}) => ({
  type: types.CHANGE_ACTIVE_SLA_OPTION,
  slaOption,
  logisticsInfoIndex,
  sellerId,
})

export const changeActiveDeliveryPackageAction = id => ({
  type: types.CHANGE_ACTIVE_DELIVERY_PACKAGE,
  id,
})

export const selectDeliveryWindowAction = (slaOption, deliveryWindow) => ({
  type: types.SELECT_DELIVERY_WINDOW,
  slaOption,
  deliveryWindow,
})

export const selectDeliveryWindow = (slaOption, deliveryWindow) => (
  dispatch,
  getState
) => {
  dispatch(selectDeliveryWindowAction(slaOption, deliveryWindow))

  const state = getState()

  const shippingData = {
    ...state.orderForm.shippingData,
    logisticsInfo: state.delivery[state.delivery.activeDeliveryOption],
  }

  dispatch(updateShippingData(shippingData, false))
}

const updateOrderFormAction = ({
  orderForm,
  address,
  searchAddress,
  channel,
}) => ({
  type: types.UPDATE_ORDERFORM,
  orderForm,
  address,
  searchAddress,
  channel,
})

export const updateSimulationOptions = (
  cheapestOption,
  fastestOption,
  combinedOption,
  selectedOption
) => ({
  type: types.UPDATE_SIMULATION_OPTIONS,
  cheapestOption,
  fastestOption,
  combinedOption,
  selectedOption,
})

export const selectAddressItemAction = selectedAddress => ({
  type: types.SELECT_ADDRESS_ITEM,
  selectedAddress,
})

export const selectAddressItem = (selectedAddress, alreadySelected = false) => (
  dispatch,
  getState
) => {
  dispatch(selectAddressItemAction(selectedAddress))

  if (alreadySelected) {
    return
  }

  const shippingData = getState().orderForm.shippingData
  dispatch(updateShippingData(shippingData))
}

export const toggleScheduledDelivery = status => ({
  type: types.TOGGLE_SCHEDULED_DELIVERY,
  status,
})

export const updateShippingDataRequest = (loading = true, shippingData) => ({
  type: types.UPDATE_SHIPPING_REQUEST,
  loading,
  shippingData,
})

export const updateShippingDataFailure = error => ({
  type: types.UPDATE_SHIPPING_FAILURE,
  error,
})

export const simulateShippingDataFailure = error => ({
  type: types.SIMULATE_SHIPPING_FAILURE,
  error,
})

export const simulateShippingDataSuccess = () => ({
  type: types.SIMULATE_SHIPPING_SUCCESS,
})

export const simulateShippingDataRequest = (loading = true) => ({
  type: types.SIMULATE_SHIPPING_REQUEST,
  loading,
})

export const removeItemsRequest = () => ({
  type: types.REMOVE_ITEMS_REQUEST,
})

// THUNKS
export const updateShippingData = (
  shippingData,
  loading,
  shouldHaveSelectedAddresses = true
) => (dispatch, getState) => {
  const state = getState()
  const canEditData = state.orderForm && state.orderForm.canEditData
  const items = getOrderFormItems(state)
  const deliveryAddress = getAddress(state)
  let searchAddress = getSearchAddress(state)
  const addressForm = state.addressForm

  searchAddress = searchAddress && {
    ...searchAddress,
    receiverName:
      canEditData && !!searchAddress ? searchAddress.receiverName : null,
  }

  const selectedAddresses = filterSelectedAddresses({
    canEditData,
    deliveryAddress,
    searchAddress,
  })

  const requestPayload = {
    logisticsInfo: getLogisticsInfoForRequest({
      addressForm,
      deliveryAddress,
      searchAddress,
      items,
      logisticsInfo: shippingData.logisticsInfo,
    }),
    ...(shouldHaveSelectedAddresses ? { selectedAddresses } : {}),
  }

  dispatch(updateShippingDataRequest(loading, requestPayload))

  return (
    window.vtexjs &&
    window.vtexjs.checkout &&
    window.vtexjs.checkout
      .sendAttachment('shippingData', requestPayload)
      .fail((response, status, error) => {
        if (
          !error ||
          !status ||
          response.statusText === 'abort' ||
          status === 'abort'
        ) {
          return
        }
        dispatch(updateShippingDataFailure(error))

        const errorObj =
          response.responseText && safelyParseJSON(response.responseText)

        const errorEventObject = {
          apiMessage: get(errorObj, 'error.message'),
          apiRoute: 'sendAttachment',
          errorMessage: error,
          errorStatus: status,
          operationId: errorObj && errorObj.operationId,
          reduxState: state && JSON.stringify(state),
          requestPayload: JSON.stringify(requestPayload),
          url: window.location.href,
        }

        $(window).trigger('checkout-ui-api-error', errorEventObject)

        dispatch(omnishippingAPIErrorEvent(errorEventObject))
      })
  )
}

export const removeItemsFromCart = ({ logisticsInfo, items }) => dispatch => {
  dispatch(removeItemsRequest())

  if (logisticsInfo) {
    return (
      get(window, 'vtexjs.checkout') &&
      window.vtexjs.checkout.removeItems(
        logisticsInfo.map(li => ({
          index: li.itemIndex,
          quantity: 0,
        }))
      )
    )
  }

  if (items) {
    return (
      get(window, 'vtexjs.checkout') &&
      window.vtexjs.checkout.removeItems(
        items.map(li => ({
          index: li.itemIndex,
          quantity: 0,
        }))
      )
    )
  }
}

export const changeActiveTab = ({
  channel,
  shouldUpdate = true,
  loading = true,
}) => (dispatch, getState) => {
  const state = getState()
  const address = getAddress(state)
  const searchAddress = getSearchAddress(state)

  dispatch(changeActiveTabAction({ channel, address, searchAddress }))

  if (!state.componentActivity.isOmniShippingCompleted && shouldUpdate) {
    const state = getState()
    const activeDeliveryPackage = getSelectedLeanOption(state)

    const isDeliveryAddressPostalCodeValid =
      address && address.postalCode && address.postalCode.valid

    const isSearchAddressPostalCodeValid =
      searchAddress &&
      searchAddress.postalCode &&
      searchAddress.postalCode.value &&
      searchAddress.postalCode.value.length > 0

    const isSearchAddressHasGeocoordinates =
      searchAddress &&
      searchAddress.geoCoordinates &&
      searchAddress.geoCoordinates.value &&
      searchAddress.geoCoordinates.value.length > 0

    const shouldHaveShippingData =
      (isDelivery(channel)
        ? isDeliveryAddressPostalCodeValid
        : isSearchAddressHasGeocoordinates || isSearchAddressPostalCodeValid) ||
      !state.orderForm.canEditData

    dispatch(
      updateShippingData(
        {
          ...(shouldHaveShippingData ? state.orderForm.shippingData : {}),
          logisticsInfo: activeDeliveryPackage,
        },
        loading,
        shouldHaveShippingData
      )
    )
  }
}

export const changeActiveSLAOption = ({
  slaOption,
  logisticsInfoIndex,
  sellerId,
}) => (dispatch, getState) => {
  dispatch(
    changeActiveSLAOptionAction({
      slaOption,
      logisticsInfoIndex,
      sellerId,
    })
  )
  const state = getState()

  const shippingData = {
    ...state.orderForm.shippingData,
    logisticsInfo: getSelectedLeanOption(state),
  }

  dispatch(updateShippingData(shippingData))
}

export const changeActiveDeliveryPackage = id => dispatch => {
  dispatch(
    setAditionalShippingData({
      selectedLeanShippingOption: id,
    })
  )

  setItem('activeDeliveryOption', id)

  dispatch(changeActiveDeliveryPackageAction(id))
}

function checkIfAllAddressesOk(addresses) {
  return (
    addresses.filter(address => {
      return address.postalCode || address.geoCoordinates.length
    }).length === addresses.length
  )
}

export const simulateShippingData = (
  logisticsInfo,
  loading,
  shouldSimulateRequest = true
) => (dispatch, getState) => {
  if (!logisticsInfo) return

  const state = getState()
  const deliveryAddress = getAddress(state)
  const searchAddress = getSearchAddress(state)
  const items = getOrderFormItems(state)
  const hasNoItems = items.length === 0

  if (
    !checkIfAllAddressesOk(state.orderForm.shippingData.selectedAddresses) ||
    hasNoItems ||
    state.componentActivity.isOmniShippingCompleted
  ) {
    return
  }

  if (shouldSimulateRequest) {
    dispatch(simulateShippingDataRequest(loading))
  }

  const localLogisticsInfo = getLogisticsInfoForRequest({
    items,
    logisticsInfo,
    deliveryAddress,
    searchAddress,
    isSimulation: true,
  })

  const selectedAddresses = filterSelectedAddresses({
    canEditData: state.orderForm.canEditData,
    deliveryAddress,
    searchAddress,
  })

  const dataRequest = {
    shippingData: {
      logisticsInfo: logisticsInfo ? localLogisticsInfo : [],
      ...(state.orderForm.canEditData ? { selectedAddresses } : {}),
    },
    orderFormId: state.orderForm.orderFormId,
  }

  const salesChannel = state.orderForm.salesChannel || '1'

  return $.ajax({
    url: `/api/checkout/pub/orderForms/simulation?sc=${salesChannel}`,
    type: 'POST',
    contentType: 'application/json; charset=utf-8',
    dataType: 'json',
    data: JSON.stringify(dataRequest),
    error(response, status, error) {
      if (!error) return

      const errorObj =
        response.responseText && safelyParseJSON(response.responseText)

      const errorEventObject = {
        apiMessage: get(errorObj, 'error.message'),
        apiRoute: 'simulation',
        errorMessage: error,
        errorStatus: response.status,
        operationId: errorObj && errorObj.operationId,
        reduxState: state && JSON.stringify(state),
        requestPayload: JSON.stringify(dataRequest),
        url: window.location.href,
      }

      $(window).trigger('checkout-ui-api-error', errorEventObject)

      dispatch(omnishippingAPIErrorEvent(errorEventObject))
    },
  })
}

function isLogisticsInfoTheSameAsActiveOption(getState, orderForm) {
  const { delivery } = getState()
  const activeOptionLogisticsInfo = delivery[delivery.activeDeliveryOption]

  if (activeOptionLogisticsInfo && activeOptionLogisticsInfo.length > 0) {
    const activeOptionLogisticsInfoUpdated = activeOptionLogisticsInfo.map(
      option => omit(option, ['quantity', 'deliveryWindow'])
    )

    return isEqual(
      sortSlasById(orderForm.shippingData.logisticsInfo),
      sortSlasById(activeOptionLogisticsInfoUpdated)
    )
  }

  return false
}

function changedCanEditData(getState, orderForm) {
  const { orderForm: stateOrderForm } = getState()
  return stateOrderForm.canEditData !== orderForm.canEditData
}

function shouldValidatePostalCode(address, state) {
  return (
    address &&
    get(address, 'postalCode.value') &&
    !state.componentActivity.isOmnishippingCompleted
  )
}

export const updateOrderForm = (
  { orderForm, channel },
  shouldSimulate = true,
  rules
) => (dispatch, getState) => {
  const state = getState()
  const delivery = state.delivery
  const searchAddress = getSearchAddress(state)
  const address = getAddress(state)
  const clientProfileIsEqual = isEqual(
    orderForm.clientProfileData,
    state.orderForm.clientProfileData
  )

  if (
    isLogisticsInfoTheSameAsActiveOption(getState, orderForm) &&
    clientProfileIsEqual
  ) {
    if (changedCanEditData(getState, orderForm)) {
      // When login is made to change data, should update orderForm
      dispatch(
        updateOrderFormAction({
          orderForm,
        })
      )
    }

    dispatch(simulateShippingDataSuccess())

    dispatch(
      updateSimulationOptions(
        delivery[CHEAPEST],
        delivery[FASTEST],
        delivery[COMBINED],
        delivery[delivery.activeDeliveryOption]
      )
    )

    const state = getState()
    const address = getAddress(state)
    if (shouldValidatePostalCode(address, state)) {
      dispatch(validatePostalCode(rules))
      dispatch(validateAddressForm(rules))
    }

    return
  }

  dispatch(
    updateOrderFormAction({
      orderForm,
      address,
      searchAddress,
      channel,
    })
  )

  const updatedState = getState()

  const logisticsInfo = get(orderForm, 'shippingData.logisticsInfo')
  if (
    logisticsInfo &&
    !!logisticsInfo.some(li => hasSLAs(li)) &&
    shouldSimulate &&
    !updatedState.componentActivity.isOmniShippingCompleted
  ) {
    simulateShippingDataOptions(updatedState, dispatch)
  }

  dispatch(validatePostalCode(rules))
  dispatch(validateAddressForm(rules))
}

function simulateShippingDataOptions(state, dispatch, loading) {
  $.when(
    state.delivery[CHEAPEST] &&
      dispatch(simulateShippingData(state.delivery[CHEAPEST], loading)),
    state.delivery[FASTEST] &&
      dispatch(simulateShippingData(state.delivery[FASTEST], loading)),
    state.delivery[COMBINED] &&
      dispatch(simulateShippingData(state.delivery[COMBINED], loading))
  )
    .done((cheapestData, fastestData, combinedData) => {
      const correctCheapest = cheapestData && cheapestData[0].logisticsInfo
      const correctFastest = fastestData && fastestData[0].logisticsInfo
      const correctCombined = combinedData && combinedData[0].logisticsInfo
      const selectedOption = state.delivery[state.delivery.activeDeliveryOption]

      dispatch(simulateShippingDataSuccess())

      if (correctCheapest || correctFastest || correctCombined) {
        dispatch(
          updateSimulationOptions(
            correctCheapest,
            correctFastest,
            correctCombined,
            selectedOption
          )
        )
      }
    })
    .fail(error => {
      dispatch(simulateShippingDataFailure(error))
    })
}

function getAddressIdForRequest(deliveryAddress, searchAddress, logisticsInfo) {
  if (isPickup(logisticsInfo)) {
    return searchAddress && searchAddress.addressId
  }

  if (isDelivery(logisticsInfo)) {
    return deliveryAddress && deliveryAddress.addressId
  }

  return null
}

export function getLogisticsInfoForRequest({
  items,
  logisticsInfo,
  deliveryAddress,
  searchAddress,
  isSimulation = false,
}) {
  const deliveryAddressWithoutValidation =
    deliveryAddress && removeValidation(deliveryAddress)
  const searchAddressWithoutValidation =
    searchAddress && removeValidation(searchAddress)

  return items.map((_, index) => {
    const logisticsInfoByIndex =
      !!logisticsInfo && logisticsInfo.find(li => li.itemIndex === index)

    const selectedSlaKey = isSimulation ? 'selectedSLAId' : 'selectedSla'

    if (logisticsInfoByIndex) {
      const deliveryWindowInfo = {
        deliveryWindow: logisticsInfoByIndex.deliveryWindow,
      }

      const addressId = getAddressIdForRequest(
        deliveryAddressWithoutValidation,
        searchAddressWithoutValidation,
        logisticsInfoByIndex
      )

      return {
        ...(!isSimulation ? deliveryWindowInfo : {}),
        ...(addressId ? { addressId } : {}),
        itemIndex: logisticsInfoByIndex.itemIndex,
        selectedDeliveryChannel: logisticsInfoByIndex.selectedDeliveryChannel,
        [selectedSlaKey]: logisticsInfoByIndex.selectedSla,
      }
    }

    return {
      addressId: deliveryAddress && deliveryAddress.addressId,
      itemIndex: index,
    }
  })
}
