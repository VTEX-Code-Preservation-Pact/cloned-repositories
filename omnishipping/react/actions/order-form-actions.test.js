import * as actions from './order-form-actions'
import * as types from './action-types'
import { PICKUP } from '../constants'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import addressFormInStore from './__mocks__/address-form-multiple-addresses.json'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('OrderForm Actions', () => {
  it('should create an action change active tab from a product', () => {
    const actionParams = {
      index: 0,
      tabOption: PICKUP,
    }

    expect(
      actions.changeActiveTab(actionParams.index, actionParams.tabOption)
    ).toMatchSnapshot()
  })

  it('should create an action change active SLAOption from a product', () => {
    expect(
      actions.changeActiveSLAOption({
        slaOption: PICKUP,
        sellerId: '1',
        logisticsInfoIndex: 0,
      })
    ).toMatchSnapshot()
  })

  it('should create an action update orderForm object', () => {
    expect(actions.updateOrderForm({ test: '' })).toMatchSnapshot()
  })

  it('should dispatch a action to selected one address', () => {
    const store = mockStore({
      address: {
        addressId: 'master',
        addressType: 'residential',
        city: 'Rio de Janeiro',
        complement: '100',
        country: 'BRA',
        geoCoordinates: [],
        neighborhood: 'Tijuca',
        number: '',
        postalCode: '333333333',
        receiverName: 'bananinha',
        reference: null,
        state: 'RJ',
        street: 'Rua Andrade Neves',
      },
    })

    store.dispatch(actions.selectAddressItemAction())

    const getActions = store.getActions()

    expect(getActions[0].type).toBe(types.SELECT_ADDRESS_ITEM)
  })

  it('should create an action to tell if orderForm is updated', () => {
    const store = mockStore({
      orderForm: {
        canEditData: true,
        shippingData: {
          logisticsInfo: [],
        },
        storePreferencesData: {
          countryCode: 'BRA',
        },
      },
    })

    const shippingData = [
      'shippingData',
      { logisticsInfo: [], selectedAddresses: [] },
    ]
    const spy = jest.spyOn(window.vtexjs.checkout, 'sendAttachment')

    return store
      .dispatch(actions.updateShippingData(shippingData[1]))
      .then(() => {
        expect(spy).toHaveBeenCalledWith(shippingData[0], shippingData[1])
      })
  })

  it('should create an action to remove items with the api', () => {
    const store = mockStore({ orderForm: [] })
    const logisticsInfo = [
      {
        itemIndex: 0,
      },
    ]
    const items = [{ index: 0, quantity: 0 }]
    const spy = jest.spyOn(window.vtexjs.checkout, 'removeItems')

    return store
      .dispatch(actions.removeItemsFromCart({ logisticsInfo }))
      .then(() => {
        expect(spy).toHaveBeenCalledWith(items)
      })
  })

  it('should select search address if item has pickup', () => {
    const mocks = {
      addressForm: addressFormInStore,
      items: [
        {
          uniqueId: '648B2D52FF0B462B9227CEFF75DA3875',
        },
      ],
      logisticsInfo: [
        {
          itemIndex: 0,
          addressId: '1533305916381',
          selectedDeliveryChannel: 'pickup-in-point',
        },
      ],
      deliveryAddress: {
        addressId: '1234',
        addressType: 'residential',
      },
      searchAddress: {
        addressId: '4321',
        addressType: 'search',
      },
      isSimulation: false,
    }

    const expected = [
      {
        deliveryWindow: undefined,
        addressId: '4321',
        itemIndex: 0,
        selectedDeliveryChannel: 'pickup-in-point',
        selectedSla: undefined,
      },
    ]

    const {
      addressForm,
      items,
      logisticsInfo,
      deliveryAddress,
      searchAddress,
      isSimulation,
    } = mocks

    expect(
      actions.getLogisticsInfoForRequest({
        addressForm,
        items,
        logisticsInfo,
        deliveryAddress,
        searchAddress,
        isSimulation,
      })[0].addressId
    ).toBe(expected[0].addressId)
  })

  it('should select residential address if item is delivery', () => {
    const mocks = {
      addressForm: addressFormInStore,
      items: [
        {
          uniqueId: '648B2D52FF0B462B9227CEFF75DA3875',
        },
      ],
      logisticsInfo: [
        {
          itemIndex: 0,
          addressId: '4585178548905',
          selectedDeliveryChannel: 'delivery',
        },
      ],
      deliveryAddress: {
        addressId: '4321',
        addressType: 'residential',
      },
      searchAddress: {
        addressId: '1234',
        addressType: 'search',
      },
      isSimulation: false,
    }

    const expected = [
      {
        deliveryWindow: undefined,
        addressId: '4321',
        itemIndex: 0,
        selectedDeliveryChannel: 'delivery',
        selectedSla: undefined,
      },
    ]

    const {
      addressForm,
      items,
      logisticsInfo,
      deliveryAddress,
      searchAddress,
      isSimulation,
    } = mocks

    expect(
      actions.getLogisticsInfoForRequest({
        addressForm,
        items,
        logisticsInfo,
        deliveryAddress,
        searchAddress,
        isSimulation,
      })[0].addressId
    ).toBe(expected[0].addressId)
  })
})
