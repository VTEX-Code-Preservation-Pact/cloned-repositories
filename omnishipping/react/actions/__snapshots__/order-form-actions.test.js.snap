// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm Actions should create an action change active SLAOption from a product 1`] = `[Function]`;

exports[`OrderForm Actions should create an action change active tab from a product 1`] = `[Function]`;

exports[`OrderForm Actions should create an action update orderForm object 1`] = `[Function]`;
